flutter/lib/src/navigation/time_to_full_display_tracker.dart

```diff
@@ -0,0 +1,117 @@
+    }
+    return _instance;
+  }
+
+  DateTime? _startTimestamp;
+  ISentrySpan? _ttfdSpan;
+  ISentrySpan? _transaction;
+  Duration _autoFinishAfter = const Duration(seconds: 30);
+
+  // End timestamp provider is only needed when the TTFD timeout is triggered
+  EndTimestampProvider _endTimestampProvider = TTIDEndTimestampProvider();
+  Completer<void> _completedTTFDTracking = Completer<void>();
+
+  Future<void> track(ISentrySpan transaction, DateTime startTimestamp) async {
+    _startTimestamp = startTimestamp;
+    _transaction = transaction as SentryTracer;
+    _ttfdSpan = transaction.startChild(SentrySpanOperations.uiTimeToFullDisplay,
+        description: '${transaction.name} full display',
+        startTimestamp: startTimestamp);
+    _ttfdSpan?.origin = SentryTraceOrigins.manualUiTimeToDisplay;
+    // Wait for TTFD to finish
+    await _completedTTFDTracking.future
+        .timeout(_autoFinishAfter, onTimeout: handleTimeout);
+
+    clear();
+  }
+
+  void handleTimeout() {
+    final ttfdSpan = _ttfdSpan;
+    final startTimestamp = _startTimestamp;
+    final endTimestamp = _endTimestampProvider.endTimestamp;
+
+    if (ttfdSpan == null ||
+        ttfdSpan.finished == true ||
+        startTimestamp == null ||
+        endTimestamp == null) {
+      _completedTTFDTracking.complete();
+      return;
+    }
+
+    _setTTFDMeasurement(startTimestamp, endTimestamp);
+    ttfdSpan.finish(
+        status: SpanStatus.deadlineExceeded(), endTimestamp: endTimestamp);
+
+    _completedTTFDTracking.complete();
+  }
+
+  Future<void> reportFullyDisplayed() async {
+    final endTimestamp = getUtcDateTime();
+    final startTimestamp = _startTimestamp;
+    final ttfdSpan = _ttfdSpan;
+
+    if (ttfdSpan?.finished == true || startTimestamp == null) {
+      _completedTTFDTracking.complete();
+      return;
+    }
+
+    _setTTFDMeasurement(startTimestamp, endTimestamp);
+    await ttfdSpan?.finish(status: SpanStatus.ok(), endTimestamp: endTimestamp);
+
+    _completedTTFDTracking.complete();
+  }
+
+  void _setTTFDMeasurement(DateTime startTimestamp, DateTime endTimestamp) {
+    final duration = endTimestamp.difference(startTimestamp);
+    final measurement = SentryMeasurement.timeToFullDisplay(duration);
+    _transaction?.setMeasurement(measurement.name, measurement.value,
+        unit: measurement.unit);
+  }
+
+  void clear() {
+    _startTimestamp = null;
+    _ttfdSpan = null;
+    _transaction = null;
+    _completedTTFDTracking = Completer();
+  }
+}
+
+/// We need to retrieve the end time stamp in case TTFD timeout is triggered.
+/// In those cases TTFD end time should match TTID end time.
+/// This provider allows us to inject endTimestamps for testing as well.
+@internal
+abstract class EndTimestampProvider {
+  DateTime? get endTimestamp;
+}
+
+@internal
+class TTIDEndTimestampProvider implements EndTimestampProvider {
+  @override
+  DateTime? get endTimestamp => TimeToInitialDisplayTracker().endTimestamp;
+}
```

Comment:
What about removing unneeded classes?
```suggestion
@internal
typedef EndTimestampProvider = DateTime? Function();

@internal
EndTimestampProvider ttidEndTimestampProvider() => () => TimeToInitialDisplayTracker().endTimestamp;

```

Reply:
Yeah looks good, didn't think of that 👍